TrainingGym/env/SumoEnvParallel.py

```diff
@@ -19,95 +22,153 @@
 
+        # If the next step of the simulation is the last step of the episode, indicate the episode is done
         if self.current_step + 1 == self.steps_per_episode:
             self.is_done = True
 
         return obs, reward, self.is_done, {}
 
-    def render(self, mode='human'):
-        # Not actually using this function
-        self.current_binary = self.guiBinary
-
-    def _next_observation(self):
+    def _next_observation(self, node):
+        # TODO: Consider a different observation function
         obs = []
         wait_counts, road_counts = self._get_road_waiting_vehicle_count()
-        # HARDCODE
-        for lane in important_roads:
-            obs.append(road_counts[lane])
-            obs.append(wait_counts[lane])
-
+        # For all important roads to this node add their vehicle and waiting vehicle counts
+        for road in node['important_roads']:
+            if road in road_counts.keys():
+                obs.append(road_counts[road])
+                obs.append(wait_counts[road])
+            else:
+                obs.append(0)
+                obs.append(0)
         return np.array(obs)
 
     def _get_reward(self):
+        # TODO: Consider a different reward function
         road_waiting_vehicles_dict, _ = self._get_road_waiting_vehicle_count()
         reward = 0.0
-
         for (road_id, num_vehicles) in road_waiting_vehicles_dict.items():
-            if road_id in important_roads:
+            if road_id in all_important_roads:
                 reward -= num_vehicles
-
         return reward
 
     def _take_action(self, action):
-        if action != controlled_lights[0]['curr_phase']:
-            controlled_lights[0]['curr_phase'] = action
-            self._set_tl_phase(controlled_lights[0]['name'], action)
+        # Change phase for learning light
+        if action != self.controlled_node['curr_phase']:
+            self.controlled_node['curr_phase'] = action
+            self._set_tl_phase(self.controlled_node['name'], action)
+
+        # Change phase for controlled lights not currently learning
+        for model in self.model_list:
+            if model['next_phase'] != model['node']['curr_phase']:
+                model['node']['curr_phase'] = model['next_phase']
+                self._set_tl_phase(model['node']['name'], model['next_phase'])
 
     def _get_road_waiting_vehicle_count(self):
-        wait_counts = {'gneE16': 0, 'gneE59': 0, 'gneE13': 0}
-        road_counts = {'gneE16': 0, 'gneE59': 0, 'gneE13': 0}
+        # TODO: Find a more efficient way of getting these values (SUMO has a batch data function that might be
+        #  interesting)
+        wait_counts = {}
+        road_counts = {}
         vehicles = self.sumo.vehicle.getIDList()
         for v in vehicles:
             road = self.sumo.vehicle.getRoadID(v)
-            if road in wait_counts.keys():
+            if road in all_important_roads:
+                if road not in wait_counts.keys():
```

Comment:
this is not super necessary for this pr, but I think we could use a `defaultdict` to avoid this initialization step.

Reply:
Maybe something we can add in a future PR